rep-2015.rst

```diff
@@ -0,0 +1,137 @@
+
+Current RCL implementation only support the ``file`` scheme.
+Furthermore, it searches these security files in an enclave subdirectory within the reserved ``enclaves`` subfolder of the root keystore, corresponding to the fully-qualified path of every enclave.
+For example, for the ``/foo/bar`` enclave, the directory structure would look like:
+
+.. code-block::
+
+    <root>
+    ├── enclaves
+    │   └── foo
+    │       └── bar
+    │           ├── cert.pem
+    │           ├── key.pem
+    │           ├── ...
+    └── public
+        ├── ...
+
+Note that it also requires the names of the files to be the ones expected.
+Specifically, all certificate and key files must have the ``.pem`` extension.
+In order to configure the security properties of a DDS participant, the path of the appropriate file in the enclave directory is added as the ``file`` URI of the corresponding property.
+For example, for the private key in the authentication plugin:
+
+.. code-block::
+
+    dds.sec.auth.private_key = file:<root>/enclaves/foo/bar/key.pem
+
+
+Terminology
+===========
+
+:Certificate:
+  A data file containing the identity and authentication information of an entity.
+
+:PKCS#11:
+  Refers to one of the Public-Key Cryptography Standards, defined in PKCS#11 Standard [2].
+
+:Keystore:
+  The keystore is the root directory where the DDS security artifacts are stored.
+  See ROS 2 Security Enclaves [3] for further information.
+
+:Enclave:
+  The process or group of processes that will share the same identity and access control rules.
+  See ROS 2 Security Enclaves [3] for further information.
+
+:URI:
+  A resource's identifier that specifies how it can be accessed.
+
+Specification
+=============
+
+Goals
+-----
+
+Support PKCS#11 URIs for certificates and key files.
+``data`` URIs are out of scope of this proposal.
+
+Proposal
+--------
+
+We want to keep the current keystore structure as much as possible, as this will enable to keep all the current implementations regarding the enclave management and the CLI features that help setting up the keystore.
+No changes should be needed to systems that do not use the PKCS#11 scheme.
+
+The problem then is how to let the RMW implementation when we want to use a ``file`` URI and when a ``pkcs11`` URI; and how to provide the values of these URIs (i.e., the file path in the case of ``file`` and the token name in the case of ``pkcs11``).
+
+This can be solved if we allow for the key and certificate files in the enclave to have ``.pem`` or ``.p11`` extensions.
+Files with ``.p11`` will contain the PKCS#11 URI instead of the actual document data.
+Then, the RMW can be aware of the file extension and set the security property accordingly.
+Taking the private key in the authentication plugin as an example:
+
+#. The RMW will look for a file with name ``key.pem`` or ``key.p11`` in the enclave.
```

Comment:
Is that possible that both files exist at the same time? if that is, `pcks11` scheme prevails to `file`?

Reply:
Yes, but only if the RMW supports it.